var PeptideModifications=function(e){return e.htmlFormatSequences=function(e){return e.data.forEach(function(e){let t=e['"Sequence"'];t=(t=(t=(t=t.replace(/{([A-Z]):(.+?)}/g,'<span class="aa_mod" data-toggle="tooltip" data-placement="top" title="$2">$1</span>')).replace(/\[(\S+?)\]/g,'<span class="aa_mod" data-toggle="tooltip" data-placement="top" title="Terminal Mod $1">&bull;</span>')).replace(/^([A-Z]){(.+?)}/g,'<span class="aa_mod" data-toggle="tooltip" data-placement="top" title="Terminal Mod $2">&bull;</span>$1')).replace(/([A-Z]){(.+?)}/g,'<span class="aa_mod" data-toggle="tooltip" data-placement="top" title="$2">$1</span>'),e['"Sequence"']=t}),e},e}(PeptideModifications||{}),ScoreSummary=function(e){return e.scoreSummary={},e.rankedScores=null,e.getRankedScores=function(){var t=[];Object.keys(e.scoreSummary).forEach(function(a){t.push([a,e.scoreSummary[a].pct_scores_present])}),t.sort(function(e,t){return t[1]-e[1]}),e.rankedScores=t,e.publish("RankedScoresAvailable",e.rankedScores)},e.generateScores=function(){var t=e.href+"/api/datasets/"+e.datasetID+"?data_type=raw_data&provider=sqlite-table&headers=True&query=",a=this;$.get(t+encodeURIComponent("SELECT score_summary.* from score_summary"),function(t){var i;t.data.forEach(function(e,t){var n={};if(0===t)i=e;else{for(var r=0;r<i.length;r++)n[i[r]]=e[r];a.scoreSummary[e[i.indexOf("score_name")]]=n}}),e.publish("ScoreSummaryComplete",e.scoreSummary)})},e.init=function(t){e.href=t.href,e.datasetID=t.datasetID,e.generateScores(),e.subscribe("RequestRankedScores",function(){e.rankedScores?e.publish("RankedScoresAvailable",e.rankedScores):e.getRankedScores()})},e}(ScoreSummary||{});function AjaxDataProvider(e){var t,a=this;this.baseQuery=e.baseQuery,this.href=e.href,this.historyID=e.historyID,this.datasetID=e.datasetID,this.rowIDField=e.rowIDField,this.columnNames=e.columnNames,this.searchColumn=e.searchColumn,this.searchTable=e.searchTable,this.tableDivID=e.tableDivID,this.filtered=!1,this.customSQL=e.customSQL,this.initCompleteCB=e.callBackFN||function(){},this.scoreSummary=e.scoreSummary||!1,this.sequenceFormatter=e.sequenceFormatter||null,this.url=this.href+"/api/datasets/"+this.datasetID+"?data_type=raw_data&provider=sqlite-table&headers=True&query=",this.recordsTotal=null,this.recordsFiltered=null,this.scoresByType={},t=a,$.get(a.url+encodeURIComponent("SELECT score_summary.score_name, score_summary.score_type FROM score_summary"),function(e){e.data.forEach(function(e){e[1]in t.scoresByType||(t.scoresByType[e[1]]=[]),t.scoresByType[e[1]].push(e[0])})}),this.buildInClause=function(e){var t=" AND "+this.searchColumn+" IN (";return t+=e+")"},this.buildLikeClause=function(e){let t=RegExp(/(LIKE \"%[A-Z]*%\")/g),a=e.match(t),i=" AND ",n=this.searchColumn;return a.forEach(function(e){i+=n+" "+e+" OR "}),i=i.slice(0,i.lastIndexOf(" OR "))},this.buildBaseQuery=function(e){var t=null;return this.customSQL?t=this.customSQL:(t="SELECT "+this.baseQuery.SELECT+" FROM "+this.baseQuery.FROM,t+=this.baseQuery.WHERE?" WHERE "+this.baseQuery.WHERE:""),e.value&&(e.value.indexOf("%")>-1?t+=this.buildLikeClause(e.value):t+=this.buildInClause(e.value)),this.customSQL||(t+=this.baseQuery.GROUPBY?" GROUP BY "+this.baseQuery.GROUPBY:""),t},this.getTotalRecordCount=function(e,t,a){var i=this.buildBaseQuery(e.search),n=this.queryData,r=e,s=t,o=a;$.get(this.url+encodeURIComponent("SELECT COUNT(*) FROM ("+i+")"),function(e){o.recordsTotal=e.data[1][0],o.recordsFiltered=o.recordsTotal,n(r,s,o)})}.bind(this),this.getFilteredRecordCount=function(e,t,a){var i=this.buildBaseQuery(e.search);i="SELECT COUNT(*) FROM ("+i+")";var n=this.queryData,r=e,s=t,o=a;$.get(this.url+encodeURIComponent(i),function(e){o.recordsFiltered=e.data[1][0],n(r,s,o)})},this.getColumnNames=function(){var e=this.columnNames;return this.rowIDField&&e.forEach(function(e){e.title===this.rowIDField&&(e.visible=!1)}.bind(this)),e},this.orderFunction=function(e){var t="";return e.forEach(function(e){var a=!1,i=this.columnNames[e.column].title;this.scoresByType.REAL.forEach(function(e){i===e&&(a=!0)}),t+=a?" ORDER BY CAST("+this.columnNames[e.column].data+" AS REAL) "+e.dir+" ":" ORDER BY "+this.columnNames[e.column].data+" "+e.dir+" "}.bind(this)),t},this.queryData=function(e,t,a){var i=a,n=i.buildBaseQuery(e.search),r=e,s=i.rowIDField,o=i.sequenceFormatter;n+=i.orderFunction(r.order),n+=" LIMIT "+r.length+" OFFSET "+r.start,$.get(i.url+encodeURIComponent(n),function(e){var a=r,i={},n=e.data[0];i.draw=a.draw,i.recordsTotal=this.recordsTotal,i.recordsFiltered=this.recordsFiltered,i.data=[],e.data.slice(1).forEach(function(e){var t={};e.forEach(function(e,a){var i,r='"'+n[a]+'"';i=e,!isNaN(parseFloat(i))&&isFinite(i)?t[r]=Number.parseFloat(e):t[r]=e}),s&&(t.DT_RowId=t['"'+s+'"'],t.DT_RowData={key:t['"'+s+'"'],fObj:t}),i.orderable=!0,i.data.push(t)}),o&&(i=o(i)),t(i)}.bind(i))},this.provideData=function(e,t){this.recordsTotal?e.search.value.length>0&&!this.filtered?(this.filtered=!0,this.getFilteredRecordCount(e,t,this)):0===e.search.value.length&&this.filtered?(this.filtered=!1,this.recordsFiltered=this.recordsTotal,this.queryData(e,t,this)):this.queryData(e,t,this):this.getTotalRecordCount(e,t,this)},this.fillDOM=function(){var e=this.initCompleteCB,t={scrollX:!0,dom:"rtipl",processing:!0,serverSide:!0,ajax:function(e,t){this.provideData(e,t)}.bind(this),columns:this.getColumnNames()};$("#"+this.tableDivID).DataTable(t).on("init.dt",function(){$("#progress-div").empty(),e()})},this.generateTable=function(){this.columnNames?this.fillDOM():$.get(this.url+encodeURIComponent(this.buildBaseQuery({value:""}))+" LIMIT 0",function(e){this.columnNames=[],e.data[0].forEach(function(e){var t={};t.data='"'+e+'"',t.title=e,this.columnNames.push(t)}.bind(this)),this.fillDOM()}.bind(this))}}var FDRPresentation=function(e){return e.ignoreScores=["theoretical mass","tic","score_name"],e.scoreName=null,e.scoreValue=null,e.softwareName=null,e.softwareVersion=null,e.fdrProtocolName=null,e.fdrProtocolValue=null,e.buildPlotlyPlots=function(){Object.keys(FDRPresentation.graphPackage).forEach(function(e,t){var a=[{y:FDRPresentation.graphPackage[e].passed,x:Array.from({length:FDRPresentation.graphPackage[e].passed.length},(e,t)=>t),mode:"markers",type:"scattergl",name:"Passed FDR Threshold",hoverinfo:"y"},{y:FDRPresentation.graphPackage[e].failed,x:Array.from({length:FDRPresentation.graphPackage[e].failed.length},(e,t)=>t),mode:"markers",type:"scattergl",name:"Failed FDR Threshold",hoverinfo:"y"}],i={title:e,xaxis:{title:"PSM Index",showticklabels:!1},height:500};document.getElementById("panel_"+t);Plotly.newPlot("panel_"+t,a,i)})},e.getScoreData=function(t){let a="SELECT PSM.passThreshold, ",i=e.href+"/api/datasets/"+e.datasetID+"?data_type=raw_data&provider=sqlite-table&headers=True&query=",n=i;t.forEach(function(e){a+='PSM."'+e+'",'}),a=a.slice(0,a.lastIndexOf(",")),a+=" FROM psm_entries PSM;",n+=encodeURIComponent(a),$.get(n,function(e){var t={},a=e.data[0].slice(1);e.data[0].slice(1).forEach(function(e){t[e]={},t[e].passed=[],t[e].failed=[]}),e.data.slice(1).forEach(function(e){e.slice(1).forEach(function(i,n){"true"===e[0]?t[a[n]].passed.push(i):t[a[n]].failed.push(i)})}),a.forEach(function(e){t[e].passed.sort(function(e,t){return e-t}),t[e].failed.sort(function(e,t){return t-e})}),FDRPresentation.graphPackage=t,n=i+encodeURIComponent("SELECT protein_detection_protocol.name,protein_detection_protocol.value,   analysis_software.name, analysis_software.version    FROM protein_detection_protocol, analysis_software"),$.get(n,function(e){FDRPresentation.fdrProtocolName=e.data[1][0],FDRPresentation.fdrProtocolValue=e.data[1][1],FDRPresentation.softwareName=e.data[1][2],FDRPresentation.softwareVersion=e.data[1][3],FDRPresentation.preparePlotlyPanel(),FDRPresentation.buildPlotlyPlots(),$("#"+FDRPresentation.divID).hide(),FDRPresentation.publish("FDRDataPrepared")})})},e.getScores=function(){let t=e.href+"/api/datasets/"+e.datasetID+"?data_type=raw_data&provider=sqlite-table&headers=True&query=";t+=encodeURIComponent('SELECT score_summary.score_name FROM score_summary WHERE score_summary.pct_scores_present = 1 AND \nscore_summary.score_type = "REAL"'),$.get(t,function(e){var t=[];e.data.slice(1).forEach(function(e){-1===FDRPresentation.ignoreScores.indexOf(e[0])&&t.push(e[0])}),FDRPresentation.getScoreData(t)})},e.preparePlotlyPanel=function(){let t='<div class="panel panel-primary"><div class="panel-heading"><h3 class="panel-title"><a role="button" data-toggle="collapse" data-parent="#accordion" href="#collapseOne" aria-expanded="true" aria-controls="collapseOne">ID Scores</a></h3></div><div id="collapseOne" class="panel-collapse collapse in" role="tabpanel" aria-labelledby="headingOne"><div class="panel-body"><div class="row"><div class="col-md-12"><h5>##SOFTWARE_NAME## (##SOFTWARE_VERSION##) ##FDR_PROTOCOL_NAME## at ##FDR_PROTOCOL_VALUE##</h5></div></div>##SCORE_DOM##</div>';var a="";Object.keys(FDRPresentation.graphPackage).forEach(function(e,t){a+='<div class="row"><div class="col-md-12" id="filter_'+t+'"></div></div>',a+='<div id="panel_'+t+'"></div>'}),t=(t=(t=(t=(t=t.replace("##SCORE_DOM##",a)).replace("##SOFTWARE_NAME##",e.softwareName)).replace("##SOFTWARE_VERSION##",e.softwareVersion)).replace("##FDR_PROTOCOL_NAME##",e.fdrProtocolName)).replace("##FDR_PROTOCOL_VALUE##",e.fdrProtocolValue),$("#"+e.divID).append($.parseHTML(t))},e.init=function(t){e.href=t.href,e.datasetID=t.datasetID,e.divID=t.divID,e.callBackFN=t.callBackFN,e.getScores()},e}(FDRPresentation||{}),IGVTrackManager=function(e){return e.validTrackTypes=["bed","gff","gff3","gtf","wig","bigWig","bedGraph","bam","vcf","seg"],e.trackGroups={bed:"annotation",gff:"annotation",gff3:"annotation",gtf:"annotation",wig:"wig",bigWig:"wig",bedGraph:"wig",bam:"alignment",vcf:"variant",seg:"seg"},e.galaxyTrackFiles=null,e.queryGalaxyHistory=function(){let t=e.galaxyConfiguration.href+"/api/histories/"+e.galaxyConfiguration.historyID+"/contents/";$.get(t,function(t){let a=[];t.forEach(function(t){let i={};Object.keys(t).indexOf("extension")>-1&&t.visible&&IGVTrackManager.validTrackTypes.indexOf(t.extension.toLowerCase())>-1&&(i.id=t.id,i.name=t.name,i.sourceType=t.extension,i.trackGroup=e.trackGroups[t.extension.toLowerCase()],"bam"===t.extension.toLowerCase()&&(i.indexURL=e.galaxyConfiguration.href+t.url+"/metadata_file?metadata_file=bam_index"),a.push(i))}),e.galaxyTrackFiles=a,e.publish("ValidTrackFilesAvailable",e.galaxyTrackFiles)})},e.init=function(t){e.galaxyConfiguration=t.galaxyConfiguration,e.subscribe("NeedValidTrackFiles",function(){e.galaxyTrackFiles?e.publish("ValidTrackFilesAvailable",e.galaxyTrackFiles):e.queryGalaxyHistory()})},e}(IGVTrackManager||{});const GenomeWarningText={text:'<p class="lead">IGV - Galaxy - Genomes</p><p>The IGV viewer requires access to full genome sequences.</p><p>Galaxy can provide access to a genome sequences via the internal API. This allows for the use of custom reference sequences.</p><p>In addition, IGV has a set of reference genomes that are available to all users of the IGV.js tool</p><p>However, the genome associated with the mz.SQLite history entry is not working with the Galaxy API nor is it a standard IGV genome.</p><p>Until this issue is resolved, you will not be able to see a reference sequence in the MVP viewer.</p>'};function IGVModule(e){this.addTrackCB=e.addTrackCB,this.igvDiv=e.igvDiv,this.data=e.data,e.genome?this.genome=e.genome:this.genome=e.dbkey||"hg19",this.hidden=!1,e.fasta_file&&(this.fasta_file=e.fasta_file),e.fasta_index&&(this.fasta_index=e.fasta_index)}IGVModule.prototype.fillChrome=function(){const e={text:'<p class="lead">Purpose</p><p>The IGV panel is the Broad Institutes <a href="http://software.broadinstitute.org/software/igv/" target="_blank">IGV</a> viewer for web applications</p><p>At the start, the IGV viewer is centered around the chromosome location you clicked on in the Peptide-Protein Viewer</p><p>You can move about the entire genome from the IGV viewer. In addition you can load tracks based on data files in your current Galaxy history (for instance: BAM, BED or GTF files).</p><p>Note that you cannot change the underlying genome. The genome is set by your Galaxy history.</p>'};let t='<div class="panel panel-default"><div class="panel-heading"><div class="row"><div class="col-md-1"><span class="genome-id-value">Genome: '+this.genome+'</span></div><div class="col-md-8"><div class="btn-group" role="group" aria-label="..."><button id="add-track" type="button" class="btn btn-default">Add Track</button></div></div><div class="col-md-1"><button class="kill-igv-browser">Hide</button><span id="igv_overview_help" class="glyphicon glyphicon-question-sign" style="padding: 5px"></span></div></div></div></div><div class="panel-body"><div id="browser-location"></div></div></div>',a=this;$("#"+this.igvDiv).append($.parseHTML(t)),$("#add-track").on("click",function(){a.addTrackCB()}),$("#igv_overview_help").on("click",function(){BuildHelpPanel.showHelp({helpText:e.text,title:"IGV Overview Help"})}),$(".kill-igv-browser").on("click",function(){a.hidden=!0,$("#igvDiv").hide()})},IGVModule.prototype.loadTrack=function(e){igv.browser.loadTrack(e)},IGVModule.prototype.goToLocation=function(e){this.hidden&&$("#igvDiv").show(),igv.browser.search(e)},IGVModule.prototype.showBrowser=function(){let e={};this.fasta_file?(e.fastaURL=this.fasta_file,e.indexURL=this.fasta_index):e.genome=this.genome;let t={reference:e,locus:this.data.chrom+":"+this.data.start+"-"+this.data.end};this.fillChrome(),igv.createBrowser(document.getElementById("browser-location"),t)};var IGVManager=function(e){let t=null;return e.goToLocation=function(t){e.browser.goToLocation(t)},e.showTrackModal=function(){let a='<div class="modal fade" id="galaxy-tracks" tabindex="-1" data-backdrop="false" role="dialog"><div class="modal-dialog" role="document"><div class="modal-content"> <div class="modal-header"><button type="button" class="close" data-dismiss="modal" aria-label="Close"><span aria-hidden="true">&times;</span></button>  <h4 class="modal-title">Load IGCV Tracks <h4></div> <div class="modal-body">  <p>Available IGV Tracks:</p>  <ul class="list-group">##TRACK_LIST##</ul></div> <div class="modal-footer">  <button type="button" class="btn btn-default" data-dismiss="modal">Close</button>  <button id="btn-load-track" type="button" class="btn btn-primary">Load Track</button> </div></div></div></div>',i="";0==t.length?($("#igvDiv").prepend('<div class="alert alert-danger" role="alert">Sorry, your Galaxy history does not contain a valid IGV track. There is nothing available to load.</div>'),$("#add-track").attr("disabled","disabled")):(t.forEach(function(e){i+="<li ",e.indexURL&&(i+=' indexURL="'+e.indexURL+'"'),i+=' trackGroup="'+e.trackGroup+'" sourcetype="'+e.sourceType+'" gid="'+e.id+'" class="list-group-item track-id-item">'+e.name+"</>"}),a=a.replace("##TRACK_LIST##",i),$("#master_modal").empty().append(a),$(".track-id-item").on("click",function(){$(this).toggleClass("selected")}),$("#btn-load-track").on("click",function(){$(".track-id-item.selected").each(function(){let t=e.galaxyConfiguration.href+"/api/histories/"+e.galaxyConfiguration.historyID+"/contents/"+$(this).attr("gid")+"/display",a={};a.type=$(this).attr("trackGroup"),a.sourceType=$(this).attr("sourcetype"),a.url=t,a.name=$(this).text(),$(this).attr("indexURL")&&(a.indexURL=$(this).attr("indexURL")),e.browser.loadTrack(a)}),$("#galaxy-tracks").modal("hide")}),$("#galaxy-tracks").modal({backdrop:!1}))},e.addIGVTrack=function(){t?e.showTrackModal(t):(e.subscribe("ValidTrackFilesAvailable",function(a){t=a,e.showTrackModal(t)}),e.publish("NeedValidTrackFiles"))},e.createNewBrowser=function(t){t.addTrackCB=e.addIGVTrack,e.browser=new IGVModule(t),e.browser.showBrowser()},e.defaultGenomeConfig=function(t){let a=["hg38","hg19","hg18","mm10","gorGor4","panTro4","panPan2","susScr11","bosTau8","canFam3","rn6","danRer11","danRer10","dm6","ce11","sacCer3"],i=a.indexOf(t.genome.toLowerCase()),n="";if(i>-1)n=a[i],t.igvConfObj.fasta_file=null,t.igvConfObj.fasta_index=null,t.igvConfObj.genome=n,e.createNewBrowser(t.igvConfObj);else{let i=/\D+/,n=i.exec(t.genome.toLowerCase());for(let r=0;r<a.length;r++){let s=i.exec(a[r]);s[0]===n[0]&&(t.igvConfObj.fasta_file=null,t.igvConfObj.fasta_index=null,t.igvConfObj.genome=s.input,e.createNewBrowser(t.igvConfObj))}BuildHelpPanel.showHelp({helpText:GenomeWarningText.text,title:"Missing Genome: "+t.genome})}},e.buildModule=function(t){let a=this.galaxyConfiguration.href+"/api/genomes/"+this.galaxyConfiguration.dbkey+"/genome_uri",i=t;fetch(a).then(e=>e.json()).then(function(t){i.fasta_file=t.fasta_file,i.fasta_index=t.fasta_index,e.createNewBrowser(i)}).catch(function(t){e.defaultGenomeConfig({genome:i.dbkey,igvConfObj:i})})},e.init=function(t){e.galaxyConfiguration=t.galaxyConfiguration},e}(IGVManager||{});let PSMProteinViewer=function(){return function e(t){var a=this,i=t;this.dbkey=t.dbkey,this.msmsRender=t.msmsRender,this.proteinObj=t.protein,this.peptideList=t.peptideList,t.genome?this.genomeList=t.genome:this.genomeList=[],this.aaAbrev={A:"ala",R:"arg",N:"asn",D:"asp",C:"cys",Q:"gln",E:"glu",G:"gly",H:"his",I:"ile",L:"leu",K:"lys",M:"met",F:"phe",P:"pro",S:"ser",T:"thr",W:"trp",Y:"tyr",V:"val",B:"asx",Z:"glx",X:"xaa"},this.modifications=[];var n=this.modifications;this.peptideList.forEach(function(e){e.hasOwnProperty("mods")&&e.mods.forEach(function(e){-1===n.indexOf(e[1])&&n.push(e[1])})}),this.prtSz=20,this.rXrY=5,this.rectOpacity=.75,this.txtYBuffer=4,this.prtSeqXPos=0,this.offsetModal=10,this.maxFocusY=0,this.ctxPrtSz=5,this.psmExclusions=[],this.margin={left:50,right:50,top:50,bottom:50},this.baseDiv=t.baseDiv||"body",this.igvDiv=t.igvDiv,$(a.baseDiv).append($.parseHTML('<div id="prt_name" class="col-md-12"></div>')),$(a.baseDiv).append($.parseHTML('<div id="genome" class="col-md-12"></div>')),$(a.baseDiv).append($.parseHTML('<div id="focus" class="col-md-12"></div>')),$(a.baseDiv).append($.parseHTML('<div id="context" class="col-md-12"></div>')),this.prtNameSVG=d3.select("#prt_name").append("svg").attr("width",function(){return $("div#prt_name").width()-a.margin.left-a.margin.right}).attr("height",50),this.genomeSVG=d3.select("#genome").append("svg").attr("width",function(){return $("div#prt_name").width()-a.margin.left-a.margin.right}).attr("height",50),this.focusSVG=d3.select("#focus").append("svg").attr("width",function(){return $("div#focus").width()-a.margin.left-a.margin.right}).attr("height","100%"),this.focusGroup=null,this.contextSVG=d3.select("#context").append("svg").attr("width",function(){return $("div#context").width()-a.margin.left-a.margin.right}).attr("height","100%"),this.contextGroup=null,this.genomeGroup=null,this.ctxXScale=d3.scaleLinear().domain([0,a.proteinObj.sequence.length]).range([0,parseInt(a.contextSVG.style("width"))]),this.lineRelocator=function(){var e=a.prtSeqXPos/a.prtSz*-1,t=e+Math.floor(parseInt(a.focusSVG.style("width"))/a.prtSz);d3.select("line.locator").attr("x1",a.ctxXScale(e)).attr("x2",a.ctxXScale(t))},this.calcExclusionZone=function(e,t,i){var n=2;return a.psmExclusions.map(function(a){(t<=a[1]&&t>=a[0]||t+(e-1)<=a[1]&&t+(e-1)>=a[0])&&(n+=1)}),a.psmExclusions.push([t,t+(e-1)]),n*i},this.styleMods=function(e,t){var a=t;e.hasOwnProperty("mods")&&e.mods.forEach(function(e){var t=0===e[0]?0:e[0]-1;$("rect.peptide_"+a+":eq("+t+")").addClass("ptm"),$("text.peptide_"+a+":eq("+t+")").addClass(e[1].toLowerCase())})},this.ttDiv=d3.select("body").append("div").attr("class","tooltip").attr("id","d3-tooltip").style("opacity",0),this.renderSVG=function(){a.prtNameSVG.append("g").append("text").text(function(){return a.proteinObj.name}).attr("x",0).attr("y",20).attr("font-family","sans-serif").attr("font-size","20px"),a.focusGroup=a.focusSVG.append("g").attr("transform","translate(0,5)").call(d3.drag().on("drag",function(){a.prtSeqXPos+=d3.event.dx,a.prtSeqXPos>0&&(a.prtSeqXPos=0),Math.abs(a.prtSeqXPos)>(a.proteinObj.sequence.length-20)*a.prtSz&&(a.prtSeqXPos=(a.proteinObj.sequence.length-20)*a.prtSz*-1),d3.select(this).attr("transform","translate("+a.prtSeqXPos+",0)"),a.genomeGroup&&a.genomeGroup.attr("transform","translate("+a.prtSeqXPos+",0)"),a.lineRelocator()})),a.genomeGroup=a.genomeSVG.append("g").attr("transform","translate(0,0)"),a.genomeList.length>0?($("div #genome").find("g").append('<svg><defs><pattern id="genomePlus" x="0", y="0" width="8" height="8" patternUnits="userSpaceOnUse"><polygon points="0,0 0,8 4,4" style="fill:#e6550d;stroke:gray;stroke-width:1;opacity:1.0"></polygon></pattern><pattern id="genomeNeg" x="0", y="0" width="8" height="8" patternUnits="userSpaceOnUse"><polygon points="8,0 8,8 4,4" style="fill:#e6550d;stroke:gray;stroke-width:1;opacity:1.0"></polygon></pattern><pattern id="ref_align" patternUnits="userSpaceOnUse" x="0" y="0" width="10" height="10"> <line x1="0" y1="10" x2="5" y2="0" style="stroke:#9c3e0d;stroke-width:1" /> <line x1="5" y1="10" x2="10" y2="0" style="stroke:#9c3e0d;stroke-width:1" /></pattern></defs></svg>'),a.genomeGroup.selectAll("rect").data(a.genomeList).enter().append("rect").attr("x",function(e){return e.cds_start/3*a.prtSz}).attr("y",0).attr("width",function(e){return(e.cds_end/3-e.cds_start/3)*a.prtSz}).attr("height",8).attr("class","genome_line").attr("fill",function(e){return"+"===e.strand?"url(#genomePlus)":"url(#genomeNeg)"}).attr("rx",a.rXrY).attr("ry",a.rXrY).on("click",function(){let t=d3.select(this).data()[0],i={};e.igvModule?IGVManager.goToLocation(t.chrom+":"+t.start+"-"+t.end):(i.igvDiv=a.igvDiv,i.data=t,i.dbkey=a.dbkey,"?"===a.dbkey?alert("The mz.sqlite database must be associated with a reference genome"):IGVManager.buildModule(i),e.igvModule=!0)}).append("svg:title").text(function(e){return"Go to "+e.chrom+":"+e.start.toLocaleString()+"-"+e.end.toLocaleString()+" on an open IGV browser."}),a.genomeGroup.selectAll("text").data(a.genomeList).enter().append("text").text(function(e){return"Chr: "+e.chrom}).attr("x",function(e){return e.cds_start/3*a.prtSz+3}).attr("y",18).attr("font-size","16").attr("transform","translate(0,10)")):(a.genomeSVG.selectAll("line").data([1]).enter().append("line").attr("x1",0).attr("x2",function(){return $("div#genome").width()}).attr("class","genome_dummy").attr("y1",0).attr("y2",0),a.genomeSVG.append("text").text("No Genomic Coordinates Available").attr("x",10).attr("y",15).attr("font-size","18")),a.variantGroup=a.genomeGroup.append("g").attr("transform","translate(0,30)"),i.variantInformation.aligns.forEach(function(e){a.variantGroup.append("g").attr("transform","translate(0,10)").append("rect").attr("x",e[0]*a.prtSz).attr("y",0).attr("width",(e[1]-e[0])*a.prtSz).attr("height",3).attr("fill","#9c3e0d").attr("class","align_match").append("svg:title").text("Aligned with reference sequence.")}),i.variantInformation.deletions.forEach(function(e){a.variantGroup.append("polygon").attr("points",function(){var t=[e.loc*a.prtSz,a.prtSz],i=[t[0]+a.prtSz/2,0],n=[t[0]-a.prtSz/2,0];return t.toString()+" "+i.toString()+" "+n.toString()}).attr("fill","#9c3e0d").append("svg:title").text("Deletion of "+e.missing+" from reference seqeunce.")}),i.variantInformation.additions.forEach(function(e){var t=e.loc,i=e.added.length;a.variantGroup.append("polygon").attr("points",function(){var e=[t*a.prtSz,a.prtSz],n=[(t+i)*a.prtSz,a.prtSz],r=[(e[0]+n[0])/2,0];return e.toString()+" "+n.toString()+" "+r.toString()}).attr("fill","#9c3e0d").append("svg:title").text("Addition to reference sequence.")}),a.focusGroup.selectAll("rect").data(a.proteinObj.sequence).enter().append("rect").attr("class",function(e,t){var a="amino_acid";return i.variantInformation.substitutions.forEach(function(e){e.loc===t&&(a="reference_mismatch")}),a}).attr("x",function(e,t){return t*a.prtSz}).attr("y",0).attr("rx",a.rXrY).attr("ry",a.rXrY).attr("width",a.prtSz).attr("height",a.prtSz).attr("opacity",a.rectOpacity),a.focusGroup.selectAll("text.aa_residue").data(a.proteinObj.sequence).enter().append("text").attr("class","aa_residue").text(function(e){return e}).attr("x",function(e,t){return t*a.prtSz+a.prtSz/2}).attr("y",a.prtSz-a.txtYBuffer).attr("text-anchor","middle"),a.focusGroup.selectAll("text.offset").data(a.proteinObj.sequence).enter().append("text").attr("class","offset").text(function(e,t){if(t%a.offsetModal==0)return t}).attr("x",function(e,t){return t*a.prtSz+a.prtSz/2}).attr("y",2*(a.prtSz-a.txtYBuffer)).attr("text-anchor","middle"),a.contextGroup=a.contextSVG.append("g").attr("transform","translate(0,0)"),a.contextGroup.append("line").call(d3.drag().on("drag",function(){var e,t=0,i=parseInt(a.focusSVG.style("width")),n=a.ctxXScale(Math.floor(parseInt(a.focusSVG.style("width"))/a.prtSz));i-(t=d3.event.x<0?0:d3.event.x)<n&&(t=i-Math.ceil(n)),e=Math.ceil(t/a.ctxXScale(1)),d3.select(this).attr("x1",function(){return a.ctxXScale(e)}).attr("x2",function(){return a.ctxXScale(e)+a.ctxXScale(Math.floor(parseInt(a.focusSVG.style("width"))/a.prtSz))}),a.prtSeqXPos=-1*Math.floor(e*a.prtSz),a.focusGroup.attr("transform","translate("+a.prtSeqXPos+",5)"),a.genomeGroup&&a.genomeGroup.attr("transform","translate("+a.prtSeqXPos+",0)")}).on("end",function(){a.focusGroup.attr("transform","translate("+a.prtSeqXPos+",5)")})).attr("class","locator").attr("x1",0).attr("y1",10).attr("x2",function(){var e=Math.floor(parseInt(a.focusSVG.style("width"))/a.prtSz);return a.ctxXScale(e)}).attr("y2",10),a.contextGroup.selectAll("rect.ctx").data(a.proteinObj.sequence).enter().append("rect").attr("class",function(){return"ctx"}).attr("x",function(e,t){return a.ctxXScale(t)}).attr("y",0).attr("width",function(){return a.ctxXScale(1)}).attr("height",a.ctxPrtSz).on("mouseover",function(e,t){var i=$("div#context").width(),n=d3.event.pageX;a.ttDiv.html(e+"<br/> Offset: "+t),n>i/2&&(n-=parseInt(a.ttDiv.style("width"))),a.ttDiv.style("left",n+"px").style("top",d3.event.pageY-50+"px"),a.ttDiv.transition().duration(200).style("opacity",.9)}).on("mouseout",function(){a.ttDiv.transition().duration(500).style("opacity",0)}),a.peptideList.map(function(e,t){var i,n=a.calcExclusionZone(e.sequence.length,e.offset,a.prtSz),r=a.msmsRender;i=e.score?e.score:"No Score",n>a.maxFocusY&&(a.maxFocusY=n),a.focusGroup.selectAll("rect.peptide_"+t).data(e.sequence).enter().append("rect").attr("class",function(a,i){var n=e.class+" peptide_"+t;return e.mismatch.indexOf(i)>-1&&(n+=" mismatch"),n}).attr("x",function(t,i){return(i+=e.offset)*a.prtSz}).attr("y",n).attr("width",a.prtSz).attr("height",a.prtSz).attr("opacity",.7),a.focusGroup.selectAll("text.peptide_"+t).data(e.sequence).enter().append("text").attr("class","txt_pep peptide_"+t).attr("feature_type",e.class).attr("score_val",i).text(function(e){return e}).attr("x",function(t,i){return(i+=e.offset)*a.prtSz+a.prtSz/2}).attr("y",n+(a.prtSz-a.txtYBuffer)).attr("text-anchor","middle").attr("spectrum_identID",e.spectrum_identID).on("click",function(){r(d3.select(this).attr("spectrum_identID"))}).on("mouseover",function(){var e=d3.select(this).attr("class"),t=d3.event.pageX,i=$("#focus").position().top,n=$("#focus").width(),r="<em>"+("psm"===d3.select(this).attr("feature_type").toLowerCase()?"Target PSM":"PSM")+"</em><br>"+d3.select(this).attr("score_val");a.modifications.forEach(function(t){e.indexOf(t.toLowerCase())>=0&&(r+="<br><p><strong>Modification:</strong>"+t+"</p>")}),a.ttDiv.html(r),a.ttDiv.style("left",function(){return t<=n/2?n/2+"px":n/4+"px"}).style("top",i+"px"),a.ttDiv.transition().duration(200).style("opacity",.9)}).on("mouseout",function(){a.ttDiv.transition().duration(500).style("opacity",0)}),n/=4,a.contextGroup.selectAll("rect.ctx_psm_"+t).data(e.sequence).enter().append("rect").attr("class",function(a,i){var n="ctx_"+e.class+" ctx_psm_"+t;return e.mismatch.indexOf(i)>-1&&(n+=" mismatch"),n}).attr("x",function(t,i){return i+=e.offset,a.ctxXScale(i)}).attr("y",n).attr("width",a.ctxXScale(1)).attr("height",a.ctxPrtSz),a.styleMods(e,t)}),a.focusSVG.attr("height",a.maxFocusY+50),a.contextSVG.attr("height",a.maxFocusY/4+50)}}}();const PeptideOverviewHelp={text:'<p class="lead">Purpose</p><p>The Peptide Overview panel gives quick access to information about: unique peptide sequences (Sequences), displayed with any covalent modifications to specific amino acids highlighted and annotated via mouse-over, number of PSMs matched to each sequence (Spectra Count) and the number of inferred proteins containing each peptide sequence (Protein Count).  Data can be sorted be each of these columns in ascending or descending order.</p><hr><p class="lead">Actions</p><p><dl><dt>PSMs for Selected Peptides</dt><dd>Show PSMs for all selected unique peptide sequences in the Peptide Overview table. All PSMs matching to this peptide are shown, without filtering by scoring metrics.</dd><dt>PSMs filtered by Score</dt><dd>You can filter PSMs by their associated scores. You can filter just the PSMs linked with the selected peptide sequences in the table or filter for PSMs from the entire dataset. To select filtering criteria, a filtering panel will become visible.</dd><dt>Load from Galaxy</dt><dd>You can load a pre-filtered single-column tabular file from Galaxy containing peptide sequences of interest. The column must contain a peptide sequence on each line, one per line.</dd><dt>Peptide-Protein Viewer</dt><dd>Displays a peptide sequence aligned within the overall protein sequence(s) from which it is derived.  Other peptides from the dataset also derived from this protein are displayed as well.   This viewer also displays genomic coordinates coding for the protein sequence, from which the IGV viewer can be launched.</dd><dt>Render</dt><dd>Generate a Lorikeet view of a single, annotated MS/MS scan for each peptide shown in the overview table. Based on the score selected from the dropdown menu, the highest quality MS/MS spectra matching each peptide will be shown.</dd><dt>Filter</dt><dd>Filter and select peptides based on sequence information query entered in the text box.</dd></dl></p>'},PSMDetailHelp={text:'<p class="lead">Purpose</p><p>This panel contains all the details for a set of selected peptide sequences, including any scoring metrics available for the highest quality PSM for the peptide. Each column is sortable.  Click on any row, and the best scoring, annotated MS/MS spectra will be generated for viewing using the <a href="https://github.com/UWPR/Lorikeet" target="_blank">Lorikeet</a> tool.</p><hr><p class="lead">Actions</p><dl><dt>Row Click</dt><dd>Click on any peptide sequence in a row to generate the view of the annotated MS/MS spectra matched to this sequence.  Clicking on multiple rows will generate separate MS/MS visualizations.</dd></dl>'};var PeptideView=function(e){return e.baseQuery={SELECT:'spectrum_counts.ENCODED_SEQUENCE AS Sequence, spectrum_counts.SPECTRA_COUNT AS "Spectra Count", protein_counts.PROTEIN_COUNT AS "Protein Count", spectrum_counts.peptide_id',FROM:"spectrum_counts, protein_counts",WHERE:"protein_counts.SII_ID = spectrum_counts.SII_ID"},e.columnValues="",e.tableElm='<table id="data-table" class="table table-bordered" cellspacing="0" width="100%"></table>',e.visibleScores=[],e.forPSMRendering=[],e.filteringFiles=[],e.candidateFiles={},e.candidateFilesPanel=!1,e.prepareFilterSequences=function(t){t.length>0?t.indexOf("%")>-1?e.filterByLike(t.split(/\s|,|;|:/).toLocaleString()):e.filterBySequences(t.split(/\W/).toLocaleString()):console.log("User wants to filter on an empty list")},e.domEdit=function(){var t=$("#"+e.baseDiv);t.empty();t.append($.parseHTML('<div class="panel panel-default"> <div class="panel-heading"><h3 class="panel-title" style="display: inline">Peptide Overview</h3><span id="peptide_overview_help" class="glyphicon glyphicon-question-sign" style="padding: 5px"></span><span class="sr-only">Help?</span></div><div class="row"><div class="col-md-6"><div id="pep-functions" class="btn-group" role="group"><button id="btn-load-from-galaxy" type="button" class="btn btn-primary" disabled="disabled" data-toggle="tooltip" data-placement="bottom" title="Enlists datasets from Galaxy history for loading">Load from Galaxy</button><button id="btn-view-in-protein" type="button" class="btn btn-primary" disabled="disabled" data-toggle="tooltip" data-placement="bottom" title="Displays peptide hits aligned within protein sequences and genomic location of translated genes">Peptide-Protein Viewer</button><button type="button" class="btn btn-primary dropdown-toggle render-btn" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false" ><span data-toggle="tooltip" data-placement="bottom" title="Generate a single MSMS scan for each peptide in the overview table. The best MSMS will be determined by the chosen score.">Render </span><span class="caret"></span><span class="sr-only">Toggle Dropdown</span></button><ul id="score-type-ul" class="dropdown-menu"></ul></div></div><div class="col-md-1"></div><div class="col-md-5"><input class="pep-filter" size="40" type="text" placeholder="Peptide Sequences for Filtering"/><button type="button" class="pep-filter btn btn-primary" data-toggle="tooltip" data-placement="bottom" title="Filter peptides based on sequence information query">Filter</button><button type="button" class="pep-filter btn btn-primary" data-toggle="tooltip" data-placement="bottom" title="Clear query for filtering peptides">Clear</button></div></div><div class="panel-body">'+e.tableElm+'</div><div class="panel-footer"><div class="btn-group" role="group"><button type="button" id="psm-all" class="btn btn-primary" data-toggle="tooltip" data-placement="bottom" title="Show PSMs for all selected peptide sequences in the Peptide Overview table">PSMs for Selected Peptides</button><div class="btn-group" role="group"><button type="button" id="psm-filtered" class="btn btn-primary dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false" ><span data-toggle="tooltip" data-placement="bottom" title="Show PSMs for peptide sequences filtered by score">PSMs Filtered by Score</span> <span class="caret"></span><span class="sr-only">Toggle Dropdown</span></button> <ul class="dropdown-menu" id="psm-filter-ul"><li value="global"><a href="#">Filter Global Peptides</a></li><li value="current"><a href="#">Filter Peptides in Peptide Overview Table</a></li></ul></div></div></div></div>')),$(".render-btn").attr("disabled","disabled"),$("#psm-filter-ul li").on("click",function(){if("current"===$(this).attr("value")){if(0===e.forPSMRendering.length){let t=$("#data-table").DataTable();t.$("tr").toggleClass("selected-peptide"),t.$(".selected-peptide").each(function(){e.forPSMRendering.push(t.$(this).data())}),e.publish("ScoreFilteringRequest",e.forPSMRendering)}}else e.publish("GlobalScoreFilterRequest");$("html, body").animate({scrollTop:$("#score_filter_div").offset().top},1e3),$("#psm-filtered").tooltip("hide")}),$("#psm-all").on("click",function(){e.forPSMRendering.length>0&&(e.appendDetailDOM("Selected Peptides"),e.createDetailTable(e.forPSMRendering))}),$("button.pep-filter").on("click",function(){"Filter"===$(this).text()?e.prepareFilterSequences($("input.pep-filter").val()):($("input.pep-filter").val(""),$("#data-table").DataTable().search(""),$("#data-table").DataTable().draw())}),$("#peptide_overview_help").on("click",function(){BuildHelpPanel.showHelp({helpText:PeptideOverviewHelp.text,title:"Peptide Overview Help"})})},e.destroyDetailTable=function(){$("#detail_div").empty()},e.appendDetailDOM=function(t){var a=$("#detail_div");a.empty(),a.append($.parseHTML('<div class="panel panel-default"> <div class="panel-heading" style="background-color: lightblue;"><div class="row"><div class="col-md-10"><h3 class="panel-title" style="display: inline;">PSM Detail for '+t+'</h3><span id="psm_detail_help" class="glyphicon glyphicon-question-sign" style="padding: 5px"></span><span class="sr-only">Help?</span></div><div class="col-md-2"><span><button class="glyphicon glyphicon-remove kill-detail-table"></button></span></div></div></div><div><button id="pep-prot-psm_view" class="btn btn-primary">Peptide-Protein Viewer</button></div><div class="panel-body"><table id="data-detail-table" class="table table-bordered" cellspacing="0" width="100%"></table></div></div>')),$(".kill-detail-table").on("click",e.destroyDetailTable),$("#psm_detail_help").on("click",function(){BuildHelpPanel.showHelp({helpText:PSMDetailHelp.text,title:"PSM Details for Selected Unique Peptides"})}),$("#pep-prot-psm_view").on("click",function(){for(var t=[],a=$("#data-detail-table").DataTable().data(),i=0;i<a.length;i++)t.push(a[i]['"id"']);e.publish("userRequestsViewInProteins",t.toString())})},e.buildScoreQuery=function(){var t=" psm_entries.id, psm_entries.sequence, psm_entries.spectrumID, psm_entries.spectrumTitle, ";return e.visibleScores.forEach(function(e){t+='psm_entries."'+e+'",'}),t=t.slice(0,t.lastIndexOf(","))},e.createDetailTable=function(t,a){let i="";t.forEach(function(e){i+='"'+e.key+'",'});let n={tableDivID:"data-detail-table",rowIDField:"id",baseQuery:{SELECT:"psm_entries.*",FROM:" psm_entries",WHERE:"psm_entries.id in ("+(i=i.slice(0,i.lastIndexOf(",")))+") "},href:e.galaxyConfiguration.href,datasetID:e.galaxyConfiguration.datasetID,callBackFN:function(){$("#data-detail-table").find("tbody").on("click","tr",function(){$(this).toggleClass("selected-peptide"),e.publish("renderPSMForPeptide",{pkid:$(this).data().key,spectrumID:$(this).data().fObj['"spectrumID"'],spectrumTitle:$(this).data().fObj['"spectrumTitle"']})})}};0===i.length&&(n.baseQuery.WHERE=""),a&&(0===n.baseQuery.WHERE.length?n.baseQuery.WHERE+=" "+a+" ":n.baseQuery.WHERE+=" AND "+a+" "),e.visibleScores.length>0&&(n.baseQuery.SELECT=e.buildScoreQuery()),n.scoreSummary=!0,psmDetailDP=new AjaxDataProvider(n),psmDetailDP.generateTable(),$("html, body").animate({scrollTop:$("#detail_div").offset().top},1e3),$("#psm-all").tooltip("hide")},e.wireTable=function(){e.publish("dataTableInitialized",{}),$("#data-table").DataTable().on("draw",function(){e.forPSMRendering=[]}),$("#data-table tbody").on("click","tr",function(){let t=$("#data-table").DataTable();if($(this).hasClass("selected-peptide")){$(this).toggleClass("selected-peptide");let a=t.$(this).data();e.forPSMRendering.indexOf(a)>-1&&e.forPSMRendering.splice(e.forPSMRendering.indexOf(a),1)}else $(this).addClass("selected-peptide"),e.forPSMRendering.push(t.$(this).data())})},e.makeTableAndQuery=function(t){e.domEdit(),e.dataProvider=new AjaxDataProvider({baseQuery:t,rowIDField:"PEPTIDE_ID",href:e.galaxyConfiguration.href,datasetID:e.galaxyConfiguration.datasetID,tableDivID:"data-table",searchColumn:"spectrum_counts.SEQUENCE",callBackFN:e.wireTable,sequenceFormatter:PeptideModifications.htmlFormatSequences,modificationCount:e.galaxyConfiguration.tableRowCount.peptide_modifications}),e.dataProvider.generateTable()},e.clearSeachFiltering=function(){$("#data-table").DataTable().search("").draw()},e.setSearchFiltering=function(e){$("#data-table").DataTable().search(e).draw()},e.filterBySequences=function(e){let t="";PeptideView.clearSeachFiltering(),e.split(",").forEach(function(e,a){t+=a>0?',"'+e+'"':'"'+e+'"'}),PeptideView.setSearchFiltering(t)},e.filterByLike=function(e){let t="";PeptideView.clearSeachFiltering(),t+=' LIKE "'+e.split(",")[0]+'"',PeptideView.setSearchFiltering(t)},e.prepareRenderScores=function(t){var a="";$("#score-type-ul").empty(),t.forEach(function(e){var t='<li s_name="'+e+'" s_dir="ASC"><a href="#"><strong>'+e+"</strong> ASC</a></li>";t+='<li s_name="'+e+'" s_dir="DSC"><a href="#"><strong>'+e+"</strong> DSC</a></li>",a=e.indexOf("PeptideShaker")>-1?t.concat(a):a.concat(t)}),$("#score-type-ul").append($.parseHTML(a)),$("#score-type-ul li").on("click",function(){var t,a=$(this).attr("s_name"),i=$(this).attr("s_dir"),n=$("#data-table").DataTable().rows().data(),r=n.length,s=[];for(t=0;t<r;t+=1)s.push(n[t]['"PEPTIDE_ID"']);e.publish("renderBestPSM",{peptideIDs:s.toString(),scoreField:a,sortDir:i})}),$(".render-btn").removeAttr("disabled")},e.resetTable=function(){$("#data-table").DataTable().destroy(),$("#data-table").empty(),e.dataProvider=null,e.makeTableAndQuery(e.baseQuery)},e.reBuildTable=function(t,a){let i={SELECT:'DISTINCT spectrum_counts.ENCODED_SEQUENCE AS Sequence, spectrum_counts.SPECTRA_COUNT AS "Spectra Count", protein_counts.PROTEIN_COUNT AS "Protein Count", spectrum_counts.peptide_id',FROM:"spectrum_counts, protein_counts, psm_entries",WHERE:'protein_counts.SII_ID = spectrum_counts.SII_ID AND  psm_entries.id = spectrum_counts.PEPTIDE_ID AND psm_entries."'+t+'" >= '+a};$("#data-table").DataTable().destroy(),$("#data-table").empty(),e.dataProvider=null,e.makeTableAndQuery(i)},e.getPeptideFilterSequences=function(e){},e.buildFileSelectPanel=function(){let t='<div id="user-filter-list" class="panel panel-default">  <div class="panel-heading">Choose Tabular File(s) for Peptide Filtering</div>\n  <div class="panel-body">##FILES_HERE##</div>  <div class="panel-footer">  <div class="btn-group" role="group">  <button type="button" class="btn btn-primary" id="filter-with">Use for Filtering</button>  <button type="button" class="btn btn-primary" id="cancel-this">Cancel</button></div>  </div></div>',a='<div class="row">';Object.keys(e.candidateFiles).forEach(function(t){a+='<div class="col-md-11 shadow candidate-file" dID="'+t+'">'+e.candidateFiles[t].name+"</div>"}),a+="</div>",t=t.replace("##FILES_HERE##",a),$("#overview_row").prepend($.parseHTML(t)),$(".candidate-file").on("click",function(){$(this).hasClass("shadow")?($(this).removeClass("shadow"),$(this).addClass("used-for-filtering")):($(this).addClass("shadow"),$(this).removeClass("used-for-filtering"))}),$("#cancel-this").on("click",function(){$(".candidate-file.used-for-filtering").each(function(){$(this).removeClass("used-for-filtering"),$(this).addClass("shadow")}),$("#user-filter-list").toggle()}),$("#filter-with").on("click",function(){e.filteringFiles=[],$(".candidate-file.used-for-filtering").each(function(){e.filteringFiles.push(e.candidateFiles[$(this).attr("dID")])}),e.filteringFiles.length>0&&e.publish("ParseCandidateFiles",e.filteringFiles),$("#user-filter-list").toggle()}),e.candidateFilesPanel=!0},e.init=function(t){e.baseDiv=t.baseDiv,e.galaxyConfiguration=t.galaxyConfiguration,e.makeTableAndQuery(e.baseQuery),e.subscribe("UserRemovedFDRFilter",function(){e.resetTable()}),e.subscribe("VisibleScores",function(t){e.visibleScores=t,e.prepareRenderScores(t)}),e.publish("RequestVisibleScores"),e.publish("RequestRankedScores",{}),e.subscribe("CandidateFilesParsed",function(t){let a=$("#overview_div");e.filteringFiles=t.data,a.removeClass("col-md-12"),a.addClass("col-md-8"),$("#data-table").DataTable().draw(),SequenceByFile.init({pageSize:50,seqData:e.filteringFiles,elID:"overview_row",callBack:e.filterBySequences})}),e.subscribe("CandidateFilesAvailable",function(t){var a=$("#btn-load-from-galaxy"),i=$("#btn-view-in-protein");a.removeAttr("disabled"),t.forEach(function(t){e.candidateFiles[t.obj.id]=t}),a.on("click",function(){e.candidateFilesPanel?$("#user-filter-list").toggle():e.buildFileSelectPanel()}),i.removeAttr("disabled"),i.on("click",function(){for(var t=[],a=$("#data-table").DataTable().data(),i=0;i<a.length;i++)t.push(a[i]['"PEPTIDE_ID"']);e.publish("userRequestsViewInProteins",t.toString())})}),e.subscribe("UserProvidesScoreFiltering",function(t){e.appendDetailDOM("Score Filtered PSMs"),e.createDetailTable(t.data,t.fStr),document.getElementById("overview_div").setAttribute("style","display:none")}),e.subscribe("UserClearedScoreFilter",function(){document.getElementById("overview_div").setAttribute("style","display:visible"),$("#data-table").DataTable().$("tr").removeClass("selected-peptide"),e.forPSMRendering=[]}),$('[data-toggle="tooltip"]').tooltip()},e}(PeptideView||{});const LorikeetHelp={text:'<p class="lead">Purpose</p><p>The <a href="http://uwpr.github.io/Lorikeet/" target="_blank">Lorikeet</a> viewer allows users to view and explore the MS/MS spectra which results in matches to specific peptide sequences.  Annotated MS/MS are shown for peptide sequences clicked on in the PSM Detail for Selected Peptides window, or those generated using the Render button. Lorikeet allows users to select various attributes of peptide MS/MS spectra to annotate (left side) and view a tabular summary peaks matching to selected fragment ions within the spectrum.  More information on Lorikeet is available at <a href="https://github.com/UWPR/Lorikeet" target="_blank">GitHub</a></p><p>Along the top, you can delete, “thumbs up” or “thumbs down” the scan. Scans receiving a thumbs up assignment are tagged, and can be exported back to Galaxy for further analysis if desired.</p>'};function LorikeetInstance(){this.options={sequence:null,staticMods:[],variableMods:[],ntermMod:0,ctermMod:0,peaks:[],massError:.1,scanNum:null,fileName:null,charge:null,precursorMz:null,ms1peaks:null,ms1scanLabel:null,precursorPeaks:null,precursorPeakClickFn:null,zoomMs1:!1,width:750,height:450,extraPeakSeries:[],residueSpecificNeutralLosses:!1,showIonTable:!0,showViewingOptions:!0,showOptionsTable:!0,showInternalIonOption:!0,showInternalIonTable:!0,showMHIonOption:!1,showAllTable:!1,showSequenceInfo:!0}}var RenderPSM=function(e){return e.currentSpectra={},e.renderedSpectra={},e.idMapping={},e.determineModType=function(e,t){return rVal=[],e.forEach(function(e){e.modType===t&&rVal.push(e)}),rVal},e.setModTypes=function(e,t){let a=e.sequence.length,i=[];void 0!=e.modifications&&(i=e.modifications),t.staticMods=[],t.variableMods=[],t.ntermMod=0,t.ctermMod=0,i.forEach(function(e){let i=!0,n={};0===e.index&&(t.ntermMod+=e.modMass,i=!1),e.index===a+1&&(t.ctermMod+=e.modMass,i=!1),i&&("fixed"===e.modType?(n.modMass=e.modMass,n.aminoAcid=e.aminoAcid,t.staticMods.push(n)):(n.index=e.index,n.modMass=e.modMass,n.aminoAcid=e.aminoAcid,t.variableMods.push(n)))})},e.renderSpectrum=function(t){var a='<div id="#ID#" class="panel panel-info col-md-12" style="background-color: #d9edf7"><div class="panel-heading"><div class="row"><div class="col-md-10"><span class="aa aa_header">#PH#</span><span class="glyphicon glyphicon-question-sign lorikeet_help" style="padding: 5px"></span><span class="sr-only">Help?</span></div><div class="col-md-2"<div class="btn-group btn-group-xs" role="group" style="padding-bottom: 5px;"><button value="#ID#" spec_id="#SID#" type="button" class="btn btn-default delete-scan"><span class="glyphicon glyphicon-remove" aria-hidden="true"></span></button><button value="#ID#" spec_id="#SID#" type="button" class="btn btn-default verify-scan"><span class="glyphicon glyphicon-thumbs-up" aria-hidden="true"></span></button><button value="#ID#" spec_id="#SID#" type="button" class="btn btn-default unverify-scan"><span class="glyphicon glyphicon-thumbs-down" aria-hidden="true"></span></button></div></div><div class="panel-body">#PB#</div></div>',i=new LorikeetInstance,n=Object.keys(e.renderedSpectra).length+1;e.idMapping[n]=t,a=(a=(a=(a=a.replace(/#ID#/g,n)).replace(/#SID#/g,t)).replace("#PH#",e.currentSpectra[t].sequence)).replace("#PB#",'<div id="lm_'+n+'"></div>'),i.scanNum=t,i.sequence=e.currentSpectra[t].sequence,i.peaks=e.currentSpectra[t].peaks,i.width=750,i.height=450,i.showInternalIonTable=!0,i.showInternalIonOption=!0,i.showAllTable=!0,i.showOptionsTable=!0,i.labelReporters=!0,i.showMHIonOption=!0,e.setModTypes(e.currentSpectra[t],i),$("#lorikeet_zone").prepend($.parseHTML(a)),$("#lm_"+n).specview(i),$(".lorikeet_help").on("click",function(){BuildHelpPanel.showHelp({helpText:LorikeetHelp.text,title:"Lorikeet MS/MS Viewer"})}),$(".unverify-scan").on("click",function(){$("#"+$(this).val()).removeClass().addClass("panel panel-danger"),e.publish("userUnverifiedSpectrum",{scanID:$(this).attr("spec_id")})}),$(".delete-scan").on("click",function(){$("#"+$(this).val()).remove()}),$(".verify-scan").on("click",function(){$("#"+$(this).val()).removeClass().addClass("panel panel-success"),e.publish("userVerifiedSpectrum",{spectrum_title:RenderPSM.renderedSpectra[e.idMapping[$(this).val()]].Spectrum_title,spectrum_pkid:RenderPSM.renderedSpectra[e.idMapping[$(this).val()]].Spectrum_pkid,peptide_pkid:RenderPSM.renderedSpectra[e.idMapping[$(this).val()]].Peptide_pkid,sequence:RenderPSM.renderedSpectra[e.idMapping[$(this).val()]].sequence})})},e.getSequences=function(){var t="SELECT peptides.id, peptides.sequence FROM peptides WHERE peptides.id in (##PEPS##)",a="",i=e.galaxyConfiguration.href+"/api/datasets/"+e.galaxyConfiguration.datasetID+"?data_type=raw_data&provider=sqlite-table&headers=True&query=";Object.keys(e.currentSpectra).forEach(function(t,i){a+=i>0?',"'+e.currentSpectra[t].Peptide_pkid+'"':'"'+e.currentSpectra[t].Peptide_pkid+'"'}),i+=t=t.replace("##PEPS##",a),$.get(i,function(t){var a={};t.data.slice(1).forEach(function(e){a[e[0]]=e[1]}),Object.keys(e.currentSpectra).forEach(function(t){e.currentSpectra[t].sequence=a[e.currentSpectra[t].Peptide_pkid]}),Object.keys(e.currentSpectra).forEach(function(t){e.renderSpectrum(t),e.renderedSpectra[t]=e.currentSpectra[t]})})},e.getModifications=function(){var t="SELECT PM.peptide_ref, PM.location, PM.residue, PM.monoisotopicMassDelta, PM.modType FROM peptide_modifications PM WHERE PM.peptide_ref in (##PEPS##)",a="",i=e.galaxyConfiguration.href+"/api/datasets/"+e.galaxyConfiguration.datasetID+"?data_type=raw_data&provider=sqlite-table&headers=True&query=";Object.keys(e.currentSpectra).forEach(function(t,i){a+=i>0?',"'+e.currentSpectra[t].Peptide_pkid+'"':'"'+e.currentSpectra[t].Peptide_pkid+'"'}),i+=t=t.replace("##PEPS##",a),$.get(i,function(t){var a={};t.data.slice(1).forEach(function(e){var t={};t.index=e[1],t.modMass=e[3],t.aminoAcid=e[2],t.modType=e[4],e[0]in a||(a[e[0]]=[]),a[e[0]].push(t)}),Object.keys(e.currentSpectra).forEach(function(t){e.currentSpectra[t].modifications=a[e.currentSpectra[t].Peptide_pkid]}),e.getSequences()})},e.buildPeaks=function(){var t='SELECT scans.mzValues, scans.intensities, scans.spectrumID, scans.spectrumTitle FROM scans \nWHERE scans.spectrumTitle = "##SPECTRUM_TITLE##" AND scans.spectrumID = "##SPECTRUM_ID##"',a=e.galaxyConfiguration.href+"/api/datasets/"+e.galaxyConfiguration.datasetID+"?data_type=raw_data&provider=sqlite-table&headers=True&query=",i="";Object.keys(e.currentSpectra).forEach(function(a,n){i=(i=(i=n>0?i+" UNION "+t:t).replace("##SPECTRUM_TITLE##",e.currentSpectra[a].Spectrum_title)).replace("##SPECTRUM_ID##",e.currentSpectra[a].Spectrum_pkid)}),a+=i,$.get(a,function(t){t.data.slice(1).forEach(function(t){var a=JSON.parse(t[0]),i=JSON.parse(t[1]),n=e.currentSpectra[t[2]+","+t[3]].peaks=[];a.forEach(function(e,t){n.push([e,i[t]])})}),e.getModifications()})},e.beginBulkRender=function(t){var a='SELECT psm_entries.id, psm_entries.spectrumID, psm_entries.spectrumTitle, ##MM##(psm_entries."##SCORE##") AS "##SCORE##" FROM psm_entries  WHERE  psm_entries.id IN (##ID_LIST##)GROUP BY psm_entries.id',i="ASC"===t.sortDir?"MIN":"MAX",n=e.galaxyConfiguration.href+"/api/datasets/"+e.galaxyConfiguration.datasetID+"?data_type=raw_data&provider=sqlite-table&headers=True&query=",r=t.peptideIDs.split(",").map(function(e){return'"'+e+'"'}).toString();n+=a=(a=(a=a.replace("##ID_LIST##",r)).replace(/##SCORE##/g,t.scoreField)).replace("##MM##",i),$.get(n,function(t){t.data.slice(1).forEach(function(t){var a={};a.Peptide_pkid=t[0],a.Spectrum_pkid=t[1],a.Spectrum_title=t[2],e.currentSpectra[a.Spectrum_pkid+","+a.Spectrum_title]=a}),e.buildPeaks()})},e.beginSinglePSM=function(t){var a="SELECT psm_entries.id, psm_entries.spectrumID, psm_entries.spectrumTitle FROM psm_entries WHERE psm_entries.spectrumID = ##ID## AND psm_entries.spectrumTitle = ##TITLE## ",i=e.galaxyConfiguration.href+"/api/datasets/"+e.galaxyConfiguration.datasetID+"?data_type=raw_data&provider=sqlite-table&headers=True&query=";i+=a=(a=a.replace("##ID##",'"'+t.spectrumID+'"')).replace("##TITLE##",'"'+t.spectrumTitle+'"'),$.get(i,function(t){t.data.slice(1).forEach(function(t){var a={};a.Peptide_pkid=t[0],a.Spectrum_pkid=t[1],a.Spectrum_title=t[2],e.currentSpectra[a.Spectrum_pkid+","+a.Spectrum_title]=a}),e.buildPeaks()})},e.manageClearing=function(){$("#clear_scans").removeAttr("disabled").on("click",function(){$("#lorikeet_zone").children().each(function(){$(this).remove()}),$(this).attr("disabled","disabled"),$("#clear_scans").tooltip("hide")})},e.init=function(t){e.galaxyConfiguration=t.galaxyConfiguration,e.subscribe("renderPSMForPeptide",function(t){e.currentSpectra={},e.beginSinglePSM({peptideID:t.pkid,spectrumID:t.spectrumID,spectrumTitle:t.spectrumTitle}),e.manageClearing(),$("html, body").animate({scrollTop:$("#lorikeet_zone").offset().top},1e3)}),e.subscribe("renderBestPSM",function(t){e.currentSpectra={},e.beginBulkRender({peptideIDs:t.peptideIDs,scoreField:t.scoreField,sortDir:t.sortDir}),$("html, body").animate({scrollTop:$("#lorikeet_zone").offset().top},1e3),e.manageClearing()})},e}(RenderPSM||{});const ScoreFilterHelp={text:'<p class="lead">Purpose</p><p>The Scores for Filtering panel allows you to search for PSMs based on individual PSM scores. From the <em>Score</em> dropdown, you can choose one or multiple PSM scores for fitering.</p><hr><p class="lead">Actions</p><p><dl><dt>All conditions are true</dt><dd>Each chosen score filter must be true</dd><dt>Any condition is true</dt><dd>Any one score filter is true</dd><dt>Filter Now</dt><dd>Produces a table of PSMs fulfilling your score filtering.</dd></p>'};var ScoreFilterModule=function(e){return e.scoreSummary=null,e.guiDiv="score_filter_div",e.buildScoreFilterRow=function(t){var a=Math.random().toString(36).replace(/[^a-z]+/g,"");let i='<div id="'+a+'" class="row"><div class="col-md-4 sf-name"><span class="lead">'+t+'</span></div><div class="col-md-3">MIN: '+e.scoreSummary[t].min_value+" MAX: "+e.scoreSummary[t].max_value+"  </div>";return i+='<div class="col-md-1"><select class="score_filter_op"><option value="gt">&gt;</option><option value="gte">&ge;</option><option value="lt">&lt;</option><option value="lte">&le;</option></select></div>',i+='<div class="col-md-2"><input class="sf-number" type="number"></div><div class="col-md-1"><span delete-row="'+a+'" class="glyphicon glyphicon-remove filter-remove" aria-hidden="true"></span></div>'},e.prepareDOM=function(){let t='<div class="panel panel-default"><div class="panel-heading"><h3 class="panel-title" style="display: inline;">Scores for Filtering</h3><span id="score_filter_help" class="glyphicon glyphicon-question-sign" style="padding: 5px"></span><span class="sr-only">Help?</span></div><div class="panel-body"><div class="dropdown"><button class="btn btn-default dropdown-toggle" type="button" id="dropdown-score" data-toggle="dropdown">Score<span class="caret"></span></button><ul class="dropdown-menu" aria-labelledby="dropdown-score">##LIST##</ul></div><div id="score-filter-rows"></div><input type="radio" value="AND" name="q-type" class="score_query_type"><label for="all-clause">All conditions are true (AND)</label><span>&nbsp;</span><input type="radio" value="OR" name="q-type" class="score_query_type" checked><label for="any-clause">Any condition is true (OR)</label></div><div class="panel-footer"><button type="button" id="score-filter-now" class="btn btn-primary btn-sm">Filter Now</button><button type="button" id="score-filter-clear" class="btn btn-primary btn-sm">Clear Filter</button></div></div>',a="";Object.keys(e.scoreSummary).sort().forEach(function(t){"REAL"===e.scoreSummary[t].score_type&&(a+='<li class="shadow score_filter_name">'+t+"</li>")}),t=t.replace("##LIST##",a),$("#"+e.guiDiv).append($.parseHTML(t)),$(".score_filter_name").on("click",function(){let t=e.buildScoreFilterRow($(this).text());$("#score-filter-rows").append(t),$(".filter-remove").on("click",function(){document.getElementById($(this).attr("delete-row")).remove()})}),$("#score-filter-clear").on("click",function(){$("#"+e.guiDiv).empty(),e.publish("UserClearedScoreFilter")}),$("#score-filter-now").on("click",function(){let t=" ",a={gt:">",gte:">=",lt:"<",lte:"<="};$("#score-filter-rows").children().each(function(){t+='psm_entries."'+$(this).find(".sf-name").text()+'" '+a[$(this).find(".score_filter_op").val()]+" "+$(this).find(".sf-number").val()+" "+$("input[name=q-type]:checked").val()+" "}),t=t.slice(0,t.lastIndexOf(" "+$("input[name=q-type]:checked").val()+" ")),e.publish("UserProvidesScoreFiltering",{fStr:t,data:e.peptideObjs})}),$("#score_filter_help").on("click",function(){BuildHelpPanel.showHelp({helpText:ScoreFilterHelp.text,title:"Score Filter Help"})})},e.prepareScoreFiltering=function(t){e.peptideObjs=t,e.prepareDOM()},e.init=function(t){t.baseDiv&&(e.guiDiv=t.baseDiv),e.subscribe("ScoreSummaryComplete",function(t){e.scoreSummary=t}),e.subscribe("ScoreFilteringRequest",function(t){e.prepareScoreFiltering(t)}),e.subscribe("GlobalScoreFilterRequest",function(){e.peptideObjs=[],e.prepareDOM()})},e}(ScoreFilterModule||{}),SequenceByFile=function(e){return e.drawTable=function(){var t='<div class="col-md-4" id="sequence_data_div"><div class="panel panel-default"><div class="panel-heading"><div class="row"><div class="col-md-10"><div class="panel-title">Filtering Sequences</div></div><div class="col-md-2"><button class="glyphicon glyphicon-remove btn-xs clear-filter"></button></div></div></div><div class="panel-body fixed-height-panel">';e.pagedData.forEach(function(e){e[1].length>0&&(t+='<div class="row shadow filter-by" data-entries="'+e[1]+'"><div class="col-md-12"><strong>'+e[0]+'</strong></div><div class="col-md-12"><small>'+e[1][0]+"</small> to <small>"+e[1].slice(-1)+"</small></div></div>")}),t+="</div></div></div>",$("#"+e.elID).append($.parseHTML(t)),$(".filter-by").on("click",function(){$(".filter-by").each(function(){$(this).hasClass("shadow")||($(this).removeClass("used-for-filtering"),$(this).addClass("shadow"))}),$(this).removeClass("shadow"),$(this).addClass("used-for-filtering"),e.filterFunc($(this).attr("data-entries"))}),$(".clear-filter").on("click",function(){var e=$("#overview_div");$("#sequence_data_div").remove(),e.removeClass("col-md-8"),e.addClass("col-md-12"),$("#data-table").DataTable().search(""),$("#data-table").DataTable().draw()})},e.pageData=function(){e.pagedData=[],Object.keys(e.seqData).forEach(function(t){var a,i=Math.floor(e.seqData[t].length/e.pageSize);if(0===i)e.pagedData.push([t,e.seqData[t]]);else for(a=0;a<=i;a+=1)e.pagedData.push([t+" : Page "+(a+1),e.seqData[t].slice(a*e.pageSize,a*e.pageSize+e.pageSize)])}),e.drawTable()},e.init=function(t){e.pageSize=t.pageSize||50,e.seqData=t.seqData,e.elID=t.elID,e.pageData(),e.filterFunc=t.callBack},e}(SequenceByFile||{}),VerifiedScans=function(e){return e.psmEntrySQL="SELECT psm_entries.* FROM psm_entries WHERE psm_entries.spectrumID = ##ID## AND psm_entries.spectrumTitle = ##TITLE##",e.galaxyHeader=[],e.verifiedScans={},e.sendToGalaxy=function(){var t=e.galaxyHeader.join("\t")+"\n",a={"files_0|url_paste":null,dbkey:"?",file_type:"tabular","files_0|type":"upload_dataset","files_0|space_to_tab":null,"files_0|to_posix_lines":"Yes"},i={history_id:e.galaxyConfiguration.historyID,tool_id:"upload1",inputs:null};Object.keys(e.verifiedScans).forEach(function(a){e.verifiedScans[a].savedToGalaxy||e.verifiedScans[a].deleted||(t+=e.verifiedScans[a].psmEntry.join("\t")+"\n",e.verifiedScans[a].savedToGalaxy=!0)}),a["files_0|url_paste"]=t,i.inputs=JSON.stringify(a),$.ajax({url:e.galaxyConfiguration.href+"/api/tools",type:"POST",error:function(){console.log("ERROR in POSTING sequence file.")},success:function(e){console.log("ID for file is "+e.outputs[0].id),VerifiedScans.verifiedScans={},$("#scan-count-badge").text(VerifiedScans.eligibleScansCount())},data:i}),e.eligibleScansCount()},e.eligibleScansCount=function(){var t=0;return Object.keys(e.verifiedScans).forEach(function(a){e.verifiedScans[a].deleted||e.verifiedScans[a].savedToGalaxy||(t+=1)}),0===t?$("#scans-to-galaxy").attr("disabled","disabled"):$("#scans-to-galaxy").removeAttr("disabled"),t},e.removeScan=function(t){t in e.verifiedScans&&(e.verifiedScans[t].deleted=!0,$("#scan-count-badge").text(e.eligibleScansCount()))},e.addScore=function(t){var a=e.psmEntrySQL.replace("##ID##",'"'+t.split(",")[0]+'"');a=a.replace("##TITLE##",'"'+t.split(",")[1]+'"');var i=t;$.get(e.url+a,function(t){0===e.galaxyHeader.length&&(e.galaxyHeader=e.galaxyHeader.concat(t.data[0])),e.verifiedScans[i].psmEntry=t.data[1]})},e.init=function(t){e.galaxyConfiguration=t.galaxyConfiguration,e.url=e.galaxyConfiguration.href+"/api/datasets/"+e.galaxyConfiguration.datasetID+"?data_type=raw_data&provider=sqlite-table&headers=True&query=",e.subscribe("userVerifiedSpectrum",function(t){t.spectrum_pkid+","+t.spectrum_title in e.verifiedScans?e.verifiedScans[t.spectrum_pkid+","+t.spectrum_title].deleted=!1:(e.verifiedScans[t.spectrum_pkid+","+t.spectrum_title]={deleted:!1,savedToGalaxy:!1,sequence:t.sequence,peptide_pkid:t.peptide_pkid,spectrum_pkid:t.spectrum_pkid,spectrum_title:t.spectrum_title},e.addScore(t.spectrum_pkid+","+t.spectrum_title)),$("#scan-count-badge").text(e.eligibleScansCount())}),e.subscribe("userUnverifiedSpectrum",function(t){e.removeScan(t.scanID)}),$("#scans-to-galaxy").on("click",function(){e.sendToGalaxy(),$("#scans-to-galaxy").tooltip("hide")}).attr("disabled","disabled")},e}(VerifiedScans||{});const FeatureViewerHelp={title:"Peptide-Protein Viewer Help",helpText:'<p class="lead">Purpose</p><p>The Peptide-Protein viewer displays the complete amino acid sequence for any selected protein within the sequence database used for matching MS/MS data.</p><ul><li>The bottom line shows an overview map of the entire sequence, along with any PSMs matched to the sequence colored in yellow bars.  The darker orange colored bar represents the peptide of interest selected originally from the Peptide Overview page which maps to this protein.</li><li>The lines above show a zoomed in view of selected regions of the protein sequence, with peptides from any PSMs below the overall protein sequence.  Amino acids identified with post-translational modifications are shaded in gray; amino acid variants as compared to a reference are colored in brown within the overall protein sequence map.</li><li>The thinner top line above the amino acid sequence depicts the coding region on the chromosome for the protein, with arrows indicating the direction of transcription for these genomic coordinates.  Breaks in this line indicate joining points of exons composing the mature gene product.</li></ul><p class="lead">Actions</p><dl><dt>Selected regions for viewing from the overall protein sequence</dt><dd>The gray rectangular box can be slid along the linear map of the protein sequence to view any given region of the protein sequence (and any corresponding PSMs) with higher resolution.</dd><dt>Opening Integrated Genomics Viewer (IGV)</dt><dd>Clicking on the thinner top line (with arrows indicating directionality) will open up the interactive IGV viewer, which can be used to map and characterize peptide sequences of interest against transcripts and genomic coding regions.   The IGV tool opens within the same browser window</dd></dl>'};var featureViewer={modalHTML:'<div class="modal fade" id="peptide_protein" tabindex="-1" data-backdrop="false" role="dialog"><div class="modal-dialog" role="document"><div class="modal-content"><div class="modal-header"><button type="button" class="close" data-dismiss="modal" aria-label="Close"><span aria-hidden="true">&times;</span></button><h4 class="modal-title">##TITLE##</h4></div><div class="modal-body">##BODY##</div><div class="modal-footer"><button type="button" class="btn btn-default" data-dismiss="modal">Cancel</button></div> </div>\x3c!-- /.modal-content --\x3e </div>\x3c!-- /.modal-dialog --\x3e </div>\x3c!-- /.modal --\x3e',requestMSMSRender:function(e){featureViewer.publish("renderSingleMSMS",e)},getVarianceData:function(e,t,a,i){const n=t,r=a,s=e,o=i;let l=featureViewer.galaxyConfiguration.href+"/api/datasets/##genomicDataset##?data_type=raw_data&provider=sqlite-table&headers=True&query=",c='SELECT VA.cigar FROM variant_annotation VA WHERE VA.name = "'+e+'"';l=l.replace("##genomicDataset##",featureViewer.cigarDataset),$.get(l+c,function(e){let t;const a=RegExp("([0-9]+)([=|X|M|I|D])","g");let i,l,c={substitutions:[],deletions:[],additions:[],aligns:[]},d=0;if(e.data.length<2)o?o(s,n,r):featureViewer.queryFeatures(n,r);else{for(t=e.data[1][0];i=a.exec(t);){switch(l=d+Number.parseInt(i[1])-1,i[2]){case"=":case"M":c.aligns.push([d,l]);break;case"X":c.substitutions.push({loc:l})}d+=Number.parseInt(i[1])}featureViewer.variantInformation=c,o?o(s,n,r):featureViewer.queryFeatures(n,r)}})},getGenomicCoordinate:function(e,t,a,i){var n=t,r=a,s=featureViewer,o=s.galaxyConfiguration.href+"/api/datasets/##genomicDataset##?data_type=raw_data&provider=sqlite-table&headers=True&query=",l='SELECT feature_cds_map.* FROM feature_cds_map WHERE feature_cds_map.name = "'+e+'"';o=o.replace("##genomicDataset##",s.genomicDataset),$.get(o+l,function(t){var a=t.data[0];featureViewer.genomicCoordinates=[],t.data.slice(1).forEach(function(e){var t={};e.forEach(function(e,i){t[a[i]]=e}),featureViewer.genomicCoordinates.push(t)}),i?i(e,n,r):featureViewer.queryFeatures(n,r)})},formatRawData:function(e){var t={};return e.data.slice(1).forEach(function(a){var i=a[5];i in t?null!==a[3]&&t[i].mods.push([a[3],a[4]]):(t[i]={},t[i].peptidePKID=a[5],t[i].seq=a[6].split(""),t[i].start=a[0],t[i].end=a[1],t[i].isDecoy=a[2],t[i].mods=[],null!==a[3]&&t[i].mods.push([a[3],a[4]]),t[i].spectrum_identID=a[7],t[i].scores=[e.data[0].slice(10),a.slice(10)])}),t},buildFeatures:function(e){var t,a=e.sqlData,i=featureViewer.peptideList[e.peptideID],n=featureViewer.genomicCoordinates?featureViewer.genomicCoordinates:null,r={dbkey:featureViewer.dbkey,igvDiv:"igvDiv",baseDiv:"#"+featureViewer.baseDiv,genome:n,msmsRender:featureViewer.requestMSMSRender,protein:(t={},i.proteins.forEach(function(a){a.protein_pkid.toString()===e.proteinID&&(t.name=a.accession+" : "+a.description,a.sequence?t.sequence=a.sequence.split(""):t.sequence=[])}),t),peptideList:[],variantInformation:featureViewer.variantInformation},s=featureViewer.formatRawData(a);Object.keys(s).forEach(function(t){var a,i,n,o={},l=s[t];o.sequence=l.seq,o.offset=l.start-1,o.mismatch=[],o.spectrum_identID=l.spectrum_identID,a=r.protein.sequence,i=o.offset,o.sequence.forEach(function(e,t){a[t+i]!==e&&o.mismatch.push(t)}),o.score=(n="<dl>",l.scores[0].forEach(function(e,t){n+="<dt>"+e+"</dt><dd>"+l.scores[1][t]+"</dd>"}),n+"</dl>"),l.mods.length>0&&(o.mods=l.mods),l.peptidePKID.toString()===e.peptideID?o.class="psm":o.class="peptide",r.peptideList.push(o)}),$("#"+featureViewer.baseDiv).empty(),featureViewer.view=new PSMProteinViewer(r),featureViewer.view.renderSVG(),$("#protein_viewer").prepend('<button type="button" id="clear_map" class="btn btn-primary btn-sm">Clear Map</button><span id="fv_help" class="glyphicon glyphicon-question-sign" style="padding: 5px"></span>'),$("#clear_map").on("click",function(){$("#"+featureViewer.baseDiv).empty(),document.getElementById("d3-tooltip").remove()}),document.getElementById("fv_help").addEventListener("click",()=>{BuildHelpPanel.showHelp(FeatureViewerHelp)})},queryFeatures:function(e,t){var a="",i=t,n='SELECT\n  PE.start, PE.end, PE.isDecoy,\n  PM.location, PM.name, PSM.id, PSM.sequence, PSM.spectrumID, PSM.spectrumTitle, ##SCORES##\n FROM\n  peptide_evidence PE,\n  peptides P, psm_entries PSM\n  LEFT OUTER JOIN peptide_modifications PM ON P.id = PM.peptide_ref\n WHERE\n  PE.dBSequence_ref = "'+e+'" AND\n  PE.peptide_ref = P.id AND\n  PSM.id = P.id';featureViewer.visibleScores?(featureViewer.visibleScores.forEach(function(e){a+='PSM."'+e+'",'}),a=a.slice(0,a.lastIndexOf(",")),n=encodeURIComponent(n.replace("##SCORES##",a))):n=n.replace("##SCORES##","PSM.*"),featureViewer.queryFunc(n,function(t){featureViewer.buildFeatures({sqlData:t,peptideID:i,proteinID:e})})},showDOM:function(){var e=featureViewer.modalHTML.replace("##TITLE##","View Peptides in Protein"),t="",a=featureViewer.peptideList;Object.keys(a).forEach(function(e,i){t+='<div class="fv-entry col-md-12"><div class="fv-seq"><a role="button" data-toggle="collapse" href="#c_'+i+'" aria-expanded="true" aria-controls="collapseOne">'+a[e].sequence+"</a></div>",t+='<div id="c_'+i+'" class="panel-collapse collapse" role="tabpanel">',a[e].proteins.forEach(function(a){t+='<div class="fv-protein" dbs_id="'+a.protein_pkid+'" pep_id="'+e+'">',t+='<div class="col-md-12 p_accession">'+a.accession+"</div>",t+="</div>"}),t+="</div></div>"}),e=e.replace("##BODY##",t),$("#master_modal").empty().append(e),$(".fv-protein").on("click",function(){if($(this).text().indexOf("REVERSED")>-1){return $("#"+featureViewer.baseDiv).append('<div class="alert alert-warning alert-dismissible" role="alert"><button type="button" class="close" data-dismiss="alert" aria-label="Close"><span aria-hidden="true">&times;</span></button><strong>Sorry,</strong> I do not map reverse sequence, decoy proteins.</div>'),void $("#peptide_protein").modal("hide")}featureViewer.genomicDataset?featureViewer.cigarDataset?featureViewer.getGenomicCoordinate($(this).find(".p_accession").text(),$(this).attr("dbs_id"),$(this).attr("pep_id"),featureViewer.getVarianceData):featureViewer.getGenomicCoordinate($(this).find(".p_accession").text(),$(this).attr("dbs_id"),$(this).attr("pep_id"),null):featureViewer.cigarDataset?featureViewer.getVarianceData($(this).find(".p_accession").text(),$(this).attr("dbs_id"),$(this).attr("pep_id"),null):featureViewer.queryFeatures($(this).attr("dbs_id"),$(this).attr("pep_id")),$("#peptide_protein").modal("hide")}),$("#peptide_protein").find(".btn-primary").on("click",function(){$("#peptide_protein").modal("hide")}),$("#peptide_protein").modal("show")},getProteinData:function(e){var t="SELECT PE.peptide_ref, DB.id, DB.accession, DB.description, DB.sequence, DB.length  FROM db_sequence DB, peptide_evidence PE  WHERE PE.peptide_ref IN (##PEP_IDS##) AND PE.dBSequence_ref = DB.id";t=t.replace("##PEP_IDS##",e.toString()),featureViewer.queryFunc(t,function(e){e.data.slice(1).forEach(function(e){var t={};"proteins"in featureViewer.peptideList[e[0]]||(featureViewer.peptideList[e[0]].proteins=[]),t.protein_pkid=e[1],t.accession=e[2],t.description=e[3],t.length=e[5],t.sequence=e[4],featureViewer.peptideList[e[0]].proteins.push(t)}),featureViewer.showDOM()})},getPeptideSequences:function(e){var t="SELECT peptides.id, peptides.sequence FROM peptides WHERE peptides.id IN (##IDs##)",a=[];e.split(",").forEach(function(e){a.push('"'+e+'"')}),t=t.replace("##IDs##",a.toString()),featureViewer.queryFunc(t,function(e){featureViewer.peptideList={},e.data.slice(1).forEach(function(e){featureViewer.peptideList[e[0]]={},featureViewer.peptideList[e[0]].sequence=e[1]}),featureViewer.getProteinData(a)})},init:function(e){var t;featureViewer.galaxyConfiguration=e.galaxyConfiguration,featureViewer.baseDiv=e.baseDiv,featureViewer.queryFunc=e.queryFunc,featureViewer.variantInformation={substitutions:[],deletions:[],additions:[],aligns:[]},featureViewer.dbkey=e.galaxyConfiguration.dbkey,t=featureViewer.galaxyConfiguration.href+"/api/histories/"+featureViewer.galaxyConfiguration.historyID+"/contents/",$.get(t,function(e){e.forEach(function(e){"sqlite"===e.extension&&!1===e.deleted&&!0===e.visible&&$.get(featureViewer.galaxyConfiguration.href+e.url,function(e){e.peek.indexOf("feature_cds_map")>-1&&(featureViewer.genomicDataset=e.id),e.peek.indexOf("variant_annotation")>-1&&(featureViewer.cigarDataset=e.id)})})}),featureViewer.subscribe("VisibleScores",function(e){featureViewer.visibleScores=e}),featureViewer.subscribe("userRequestsViewInProteins",function(e){featureViewer.getPeptideSequences(e)})}},gQuery={href:null,dataSetID:null,query:function(e,t){var a=gQuery.href;a+="/api/datasets/"+gQuery.datasetID+"?data_type=raw_data&provider=sqlite-table&headers=True&query=",a+=e,$.get(a,function(e){t(e)}).fail(function(e){alert("ERROR: query \n"+e.responseText+" \nfailed.")})},init:function(e){gQuery.href=e.href,gQuery.datasetID=e.datasetID}},PeptideSequenceFilter=function(e){return e.pepsByFile={},e.candidateFiles=[],e.timer,e.historyContents=function(t){return new Promise(function(a,i){e.timer=setTimeout(function(){alert("Sorry, Galaxy is not responding. I have waited as long as I can. Grab a coffee, relax, and refresh the app in a bit."),i(new Error("Requesting data from Galaxy timeout"))},3e4),$.get(t,a)})},e.pepAJAX=function(t,a){var i=e.galaxyConfiguration.href+"/api/histories/"+e.galaxyConfiguration.historyID+"/contents/"+t.obj.id+"/display",n=a,r=t;$.get(i,function(t){var a=[];t.split("\n").forEach(function(e){var t=new RegExp("[^a-z]|sequence","i");e.match(t)||e.length>0&&a.push(e)}),e.pepsByFile[r.obj.name]=a,n()})},e.peptideFileContents=function(t){var a=t;return new Promise(function(t){e.pepAJAX(a,t)})},e.processHistoryList=function(t){var a=[];clearTimeout(e.timer),t.forEach(function(e){"tabular"!==e.extension||e.deleted||a.push({name:e.name,obj:e})}),e.candidateFiles=a,e.publish("CandidateFilesAvailable",e.candidateFiles),e.subscribe("ParseCandidateFiles",function(t){e.parseCandidateFiles(t)})},e.parseCandidateFiles=function(t){t.map(e.peptideFileContents).reduce(function(e,t){return e.then(function(){return t})},Promise.resolve()).then(function(){e.publish("CandidateFilesParsed",{data:e.pepsByFile})})},e.loadData=function(){var t=e.galaxyConfiguration.href+"/api/histories/"+e.galaxyConfiguration.historyID+"/contents";e.historyContents(t).then(e.processHistoryList)},e.init=function(t){e.galaxyConfiguration=t.galaxyConfiguration,e.subscribe("dataTableInitialized",function(){e.loadData()})},e}(PeptideSequenceFilter||{}),BuildHelpPanel=function(e){return e.domStr='<div id="user_help_modal" class="modal fade" tabindex="-1" role="dialog">\n  <div class="modal-dialog" role="document">\n    <div class="modal-content">\n      <div class="modal-header">\n        <button type="button" class="close" data-dismiss="modal" aria-label="Close"><span aria-hidden="true">&times;</span></button>\n        <h4 class="modal-title">#MODAL_TITLE#</h4>\n      </div>\n      <div class="modal-body">\n        <p>#HELP_TEXT#</p>\n      </div>\n      <div class="modal-footer">\n        <button type="button" class="btn btn-default" data-dismiss="modal">Close</button>\n      </div>\n    </div>\x3c!-- /.modal-content --\x3e\n  </div>\x3c!-- /.modal-dialog --\x3e\n</div>\x3c!-- /.modal --\x3e',e.finalDom="",e.showHelp=function(t){e.finalDom=e.domStr.replace("#HELP_TEXT#",t.helpText),e.finalDom=e.finalDom.replace("#MODAL_TITLE#",t.title),$("#master_modal").empty().append(e.finalDom),$("#user_help_modal").modal("show")},e}(BuildHelpPanel||{}),ConfigModal=function(e){return e.userDefaults={tool_tip_visible:!0},e.domStr='<div id="app_config_modal" class="modal fade" tabindex="-1" role="dialog">\n  <div class="modal-dialog" role="document">\n    <div class="modal-content">\n      <div class="modal-header">\n        <button type="button" class="close" data-dismiss="modal" aria-label="Close"><span aria-hidden="true">&times;</span></button>\n        <h4 class="modal-title">Configuration</h4>\n      </div>\n      <div class="modal-body">       <div>        <input type="checkbox" id="config_tooltip" class="app_config" value="tool_tip_visible"/>        <label for="config_tooltip">Enable tooltips</label>       </div></div>      <div class="modal-footer">\n        <button type="button" class="btn btn-default" data-dismiss="modal">Close</button>        <button id="save_config_change" type="button" class="btn btn-default">Save Changes</button>      </div>\n    </div>\x3c!-- /.modal-content --\x3e\n  </div>\x3c!-- /.modal-dialog --\x3e\n</div>\x3c!-- /.modal --\x3e',e.showConfig=function(){$("#master_modal").empty().append(e.domStr),Object.keys(e.userDefaults).forEach(function(t){var a='input[value="'+t+'"]';document.querySelector(a).checked=e.userDefaults[t]}),$("#save_config_change").on("click",function(){var t={};document.querySelectorAll('input[class="app_config"]').forEach(function(a){t[a.getAttribute("value")]=a.checked,e.userDefaults[a.getAttribute("value")]=a.checked}),e.publish("UserChangedDefaults",t),$("#app_config_modal").modal("hide")}),$("#app_config_modal").modal("show")},e}(ConfigModal||{}),ScoreDefaults=function(e){return e.defaultThreshold=.85,e.scoreProperties={},e.resetDOM=function(){$("#score_default_div").empty(),$("#score_defaults").removeAttr("   disabled")},e.modifyVisibleScores=function(){var t=[];$("div .score_name").each(function(){e.scoreProperties[$(this).text()]=$(this).hasClass("selected")}),Object.keys(e.scoreProperties).forEach(function(a){e.scoreProperties[a]&&t.push(a)}),e.publish("VisibleScores",t),e.resetDOM()},e.showDOM=function(){let t='<div class="col-md-12" id="score_choice_div"><div class="panel panel-default"><div class="panel-heading"><div class="row"><div class="col-md-8"><div class="panel-title"><strong>Choose Score Visibility</strong></div></div><div class="col-md-2"><button id="set_score_defaults_btn" class="btn">Set Defaults</button></div><div class="col-md-2"><button id="clear_div_btn" class="btn">Cancel</button></div></div></div><div class="panel-body fixed-height-panel col-md-12"><div class="row">';var a="";Object.keys(e.scoreProperties).forEach(function(t,i){0===i&&(a='<div class="row">'),i%2==0&&i>0&&(a+='<div class="row">'),e.scoreProperties[t]?a+='<div class="col-md-5 score_name selected"><strong>'+t+"</strong></div>":a+='<div class="col-md-5 score_name"><strong>'+t+"</strong></div>",i%2>0&&(a+="</div>")}),t+=a+"</div></div></div></div>",$("#score_default_div").append(t),$("div .score_name").on("click",function(){$(this).toggleClass("selected")}),$("#clear_div_btn").on("click",function(){e.resetDOM()}),$("#set_score_defaults_btn").on("click",e.modifyVisibleScores)},e.prepareDOM=function(){let t=$("#score_defaults");t.removeAttr("disabled"),t.on("click",function(){$("#score_defaults").attr("disabled","disabled"),e.showDOM(),$("#score_defaults").tooltip("hide")})},e.init=function(){e.subscribe("RequestVisibleScores",function(){var t=[];Object.keys(e.scoreProperties).forEach(function(a){e.scoreProperties[a]&&t.push(a)}),e.publish("VisibleScores",t),e.prepareDOM()}),e.subscribe("RankedScoresAvailable",function(t){t.forEach(function(t){t[1]>e.defaultThreshold?e.scoreProperties[t[0]]=!0:e.scoreProperties[t[0]]=!1})}),e.publish("RequestRankedScores")},e}(ScoreDefaults||{});const MVPHelp={text:'<p class="lead">Purpose</p><p>The Galaxy-based MVP visualization plugin enables viewing of results produced from workflows integrating genomic sequencing data and mass spectrometry (MS)-based proteomics data, commonly known as proteogenomics.  The multi-functional tool allows for organization and filtering of results, quality assessment of tandem mass spectrometry (MS/MS) data matched to peptide sequences, and visualization of data at both the protein and gene level.  A user can, with relatively few keystrokes, filter and order large datasets down to a manageable subset. Due to the tools use of server-side caching, large data sets are handled as quickly as small datasets.</p><p>MVP incorporates the Lorikeet viewer for visualizing MS/MS data as well as the Integrated Genomics Viewer (IGV) for mapping peptide and protein sequences to genomic coordinates, essential for characterizing protein sequence variants arising from gene and transcript variants.</p><p>The starting page provides peptide-centric overview of peptide spectral matches (PSMs) identified using sequence database searching against a protein sequence database, and summarized in a MZSqlite-formatted input file. From this page, a user can carry-out a number of operations, including:</p><ul><li>PSMs linked to unique peptide sequences of interest</li><li>PSMs passing scoring thresholds to ensure quality</li><li>Visualization of MS/MS annotation and quality for selected PSMs</li><li>Mapping PSMs to overall protein sequences</li><li>Mapping protein sequences to genomic coordinates</li><li>Visualization of protein sequence variants comparison to reference protein and gene sequences</li><li>Archiving of PSMs of interest and automated transfer of selected results back to Galaxy for further analyses</li><li>Archiving of visualizations</li></ul><hr><p class="lead">Actions</p><p><dl><dt>ID Scores</dt><dd>Clicking on this button provides a view of the distribution of PSMs for the entire dataset based on false discovery rate (FDR) estimates of the sequence database searching tools used to produce PSMs, if available.  This shows the percent of PSMs passing FDR thresholds dependent on assigned PSM scores. Clicking on the button a second time removes the graph.</dd><dt>ID Features</dt><dd>Clicking on this button generates the list of scoring metrics produced by the sequence database searching software used to generate PSMs.  The user can select the scoring metrics to be shown when examining in detail any PSMs matching to peptide sequences of interest.</dd><dt>Export Scans</dt><dd>Clicking this button automatically exports selected peptide sequences and associated MS/MS data back to Galaxy, where it is shown as a new item in the active History in tabular format</dd>'};var MVPApplication=function(e){return e.debug=!0,e.galaxyConfiguration={},e.events={},e.userDefaults=null,e.app_defaults={tool_tip_visible:!0},e.subscribe=function(t,a){return e.events[t]||(e.events[t]=[]),e.events[t].push({context:this,callback:a}),this},e.unsubscribe=function(t){e.event[t].filter(function(e){return this===e.context}.bind(this))},e.publish=function(t){var a,i;return!!e.events[t]&&(a=Array.prototype.slice.call(arguments,1),e.debug&&console.log("APP PUBLISH: "+t+" ARGS: "+a),e.events[t].map(function(e){(i=e).callback.apply(i.context,a)}),this)},e.installTo=function(t){t.subscribe=e.subscribe,t.publish=e.publish,t.unsubscribe=e.unsubscribe},e.init=function(t){this.galaxyConfiguration=t,$("#dataset_name").text(this.galaxyConfiguration.dataName),this.subscribe("ScoreSummaryComplete",function(){console.log("Score summary complete, begin table construction"),ScoreDefaults.init(),PeptideView.init({galaxyConfiguration:this.galaxyConfiguration,baseDiv:"overview_div"})}),this.subscribe("FDRDataPrepared",function(){$("#fdr_module").removeAttr("disabled").on("click",function(){$("#fdr_div").toggle(),$("html, body").animate({scrollTop:$("#fdr_div").offset().top},500),$("#fdr_module").tooltip("hide")})}),this.subscribe("UserChangedDefaults",function(t){Object.keys(t).forEach(function(a){console.log("UserChangedDefaults "+a+" : "+t[a]),e.app_defaults[a]=t[a],"tool_tip_visible"===a&&(t[a]?$('[data-toggle="tooltip"]').tooltip():$('[data-toggle="tooltip"]').tooltip("destroy"))})}),this.installTo(ScoreSummary),this.installTo(PeptideView),this.installTo(RenderPSM),this.installTo(VerifiedScans),this.installTo(PeptideSequenceFilter),this.installTo(featureViewer),this.installTo(gQuery),this.installTo(ScoreDefaults),this.installTo(FDRPresentation),this.installTo(ScoreFilterModule),this.installTo(IGVManager),this.installTo(IGVTrackManager),this.installTo(BuildHelpPanel),this.installTo(ConfigModal),ScoreFilterModule.init({}),gQuery.init({href:this.galaxyConfiguration.href,datasetID:this.galaxyConfiguration.datasetID}),featureViewer.init({galaxyConfiguration:this.galaxyConfiguration,baseDiv:"protein_viewer",queryFunc:gQuery.query,dbkey:this.galaxyConfiguration.dbkey}),RenderPSM.init({galaxyConfiguration:this.galaxyConfiguration}),VerifiedScans.init({galaxyConfiguration:this.galaxyConfiguration}),PeptideSequenceFilter.init({galaxyConfiguration:this.galaxyConfiguration}),ScoreSummary.init({href:this.galaxyConfiguration.href,datasetID:this.galaxyConfiguration.datasetID}),IGVManager.init({galaxyConfiguration:this.galaxyConfiguration}),IGVTrackManager.init({galaxyConfiguration:this.galaxyConfiguration}),"protein_detection_protocol"in t.tableRowCount&&FDRPresentation.init({href:this.galaxyConfiguration.href,datasetID:this.galaxyConfiguration.datasetID,divID:"fdr_div",callBackFN:PeptideView.reBuildTable}),$('[data-toggle="tooltip"]').tooltip(),$("#mvp_help").on("click",function(){BuildHelpPanel.showHelp({helpText:MVPHelp.text,title:"Multi-omics Visualization Platform (MVP) Viewer"})}),$("#mvp_full_window").on("click",function(){$("#mvp_full_window").tooltip("hide"),window.open("#","_blank")}),$("#mvp_config").on("click",function(){ConfigModal.showConfig()})},{run:function(t){e.init(t)}}}(MVPApplication||{});